### PR TITLE
Expanding Growth-Maturity-Decline

### PR DESCRIPTION
**Pull request by BenLloydPearson**
I expanded on the growth-maturity-decline section and organized it into three distinct areas. These metrics are based on what I've been using for my own work. I included all of the metrics that were here before except for contributor growth, since we've discussed moving that to another domain in issue #25.

The one area that I think might need some clarification are the histograms. I tried to describe them as best I can, but let me know if I need to explain these better.

### CODE REVIEW

**Comment by GeorgLink** [comment]
@BenLloydPearson thank you for this pull request.

I think the three categories are a helpful way to think about the data.
I'll leave this pull request open for others to comment on as well.

We have another pull request #29 that we would probably merge first.

**Comment by jgbarah** [comment]
@BenLloydPearson I think the categorization you propose is useful. However, I agree with @GeorgLink, I would rather finish and merge #29 before entering into details with this pr.

**Comment by jgbarah** [comment]
> The one area that I think might need some clarification are the histograms. I tried to describe them as best I can, but let me know if I need to explain these better.

As I commented in a review to the text, I think we could focus on metrics, and later we can comment on how to visualize them, maybe depending on the aims.

**Comment by GeorgLink** [comment]
@BenLloydPearson thank you again for this pull request. We completed work on #29 and caused a conflict with your work. Would you be able to resolve the conflict?

**Comment by BenLloydPearson** [comment]
@jgbarah, your point makes perfect sense and I agree. Instead of histogram, perhaps I can change it to "distribution." Such as: "Distribution of open issue time | The total occurrences of time distributions an issue remained open."

For example, X issues took 1 day to close, Y issues took 2 days to close, Z issues took 3 days to close, etc. I'm trying to think of the best way to describe this succinctly, so any help that can be provided is appreciated.

**Comment by germonprez** [comment]
@BenLloydPearson, how about: 

Distribution of open issues over time | Proportional frequency of time to close issues

**Comment by BenLloydPearson** [comment]
I updated the references to histograms as recommended in this thread.

File: 2_Growth-Maturity-Decline.md [hunk 1]
```diff
@@ -1,15 +1,33 @@
...
- Number of New Contributors | Number of first time contributors over time
- Community Growth | Number of new developers over time
- Community Toxicity | Toxic behavior and language over time
+ Open issues | Total number of open issues.
+ Closed issues | Total number of closed issues.
+ Resolution efficiency | Number of closed issues / number of abandoned issues.
+ Average & median time of open issues | The average & median amount of time issues remain open.
+ Average & median time of first response | The average & median amount of time until the first response to an issue is made.
+ Distribution of open issues over time | Proportional frequency of open issue time duration.
+ Distribution of issue resolution over time | Proportional frequency of closed issue time duration.
```
Comment (GeorgLink) [inline 2_Growth-Maturity-Decline.md]: > Distribution of open issues over time | Proportional frequency of open issue time duration.
> Distribution of issue resolution over time | Proportional frequency of closed issue time duration.

I'm not clear what these mean. Is there a different way to express these?

Do we even need them if we assume that all metrics are interesting as a time series? 
That is, could the first two criteria (open issues, closed issues) be duplicates of these last two?

File: 2_Growth-Maturity-Decline.md [hunk 2]
```diff
@@ -1,15 +1,33 @@
...
+
+## Code Development
+ Activity Metric | Description
+ --- | ---
+ Number of commits | Total number of commits.
+ Number of lines changed | Total number of lines of code that have been changed.
+ Average & median time to merge code | The average & median amount of time it takes to get code merged upstream.
```
Comment (GeorgLink) [inline 2_Growth-Maturity-Decline.md]: Could we formulate this more precise in terms of observable events, e.g. "time from patch or pull request submitted to merge"?

File: 2_Growth-Maturity-Decline.md [hunk 3]
```diff
@@ -1,15 +1,33 @@
...
+ Distribution of open issues over time | Proportional frequency of open issue time duration.
+ Distribution of issue resolution over time | Proportional frequency of closed issue time duration.
+
+## Code Development
+ Activity Metric | Description
+ --- | ---
+ Number of commits | Total number of commits.
+ Number of lines changed | Total number of lines of code that have been changed.
+ Average & median time to merge code | The average & median amount of time it takes to get code merged upstream. 
+ Distribution of time to merge | Proportional frequency of code merge to upstream time duration.
```
Comment (GeorgLink) [inline 2_Growth-Maturity-Decline.md]: Is this a duplicate of the previous criteria but looked at over a longer period of time? Can we remove this then?

File: 2_Growth-Maturity-Decline.md [hunk 4]
```diff
@@ -1,15 +1,33 @@
...
+ Review efficiency | Number of merged patches / number of abandoned patches.
```
Comment (GeorgLink) [inline 2_Growth-Maturity-Decline.md]: Is this a metric across all of time or is this ratio for a specific time window (e.g. submitted over 30 days)?

File: 2_Growth-Maturity-Decline.md [hunk 5]
```diff
@@ -1,15 +1,33 @@
...
+
+## Project Communication
+ Activity Metric | Description
+ --- | ---
+ Total communications | Total number of posts made to communication platforms.
```
Comment (GeorgLink) [inline 2_Growth-Maturity-Decline.md]: Maybe we can add to this that communication platforms are: email lists, discourse, slack, IRC, github, issue tracker, code repository, stack exchange, facebook, .... and what ever else a community might be using. I think this is implied but making it explicit avoids discussions later.

File: 2_Growth-Maturity-Decline.md [hunk 6]
```diff
@@ -1,15 +1,33 @@
...
+ Average & median communication response time | The average & median time taken for communications to be responded to.
```
Comment (GeorgLink) [inline 2_Growth-Maturity-Decline.md]: We have been hearing that responses are good, but not always helpful. For example a message of type "+1" does not help. A solution we have been hearing is to look specifically for time taken for a response from a maintainer. Would that change still get to the core idea behind this criteria?

File: 2_Growth-Maturity-Decline.md [hunk 7]
```diff
@@ -1,15 +1,33 @@
...
+ Average & median communication responses per thread | The average & median number of responses that are made to each communication thread or topic.
```
Comment (GeorgLink) [inline 2_Growth-Maturity-Decline.md]: How is this different to the previous one? Can we measure response time for messages that are not related (i.e. in the same thread) - asking for previous criteria because I imagined we had to measure on a per thread basis?

File: 2_Growth-Maturity-Decline.md [hunk 8]
```diff
@@ -1,15 +1,33 @@
...
+ Distribution of communication responses per thread | Proportional frequency of responses per communication thread.
```
Comment (GeorgLink) [inline 2_Growth-Maturity-Decline.md]: Same as previous but looked at over time? Can we remove the duplicate?

**Comment by GeorgLink** [comment]
Another thought: Should we promote this pull request on the mailing list to get feedback from folks who are not listening to changes on GitHub?

**Comment by BenLloydPearson** [comment]
@GeorgLink 

> 
> It appears that the following metrics got lost. This might be because they could not be easily considered part of one of the three metric categories. Mabe a new catory could be "Community" with the following metrics:
> 
>     Number of New Contributors
>     Community Growth (note, we might want to change that to: recruited and retained developers - include leaving developers in addition to new developers)
>     Community Toxicity

As I mentioned in my original post, in issue #25 we discussed moving these metrics into a different category so that these metrics can focus on the software itself, while Diversity & Participation focuses on the community side of things. Once we've finalized this PR, I was planning on expanding the Diversity section in a similar way as this.

> Another thought: Should we promote this pull request on the mailing list to get feedback from folks who are not listening to changes on GitHub?

I posted to the mailing list when I originally submitted this PR, but I'd have no problem reminding people, especially considering some might have been traveling to the OS Summit in the last week. Otherwise, I'll take a closer look at your comments and make changes.

File: 2_Growth-Maturity-Decline.md (hunk 1 above)
Comment (BenLloydPearson) [reply 2_Growth-Maturity-Decline.md]: There are slight differences. Average and median cover the collective total of all issues, whereas the distribution shows you each individual item. For example, given 10 items, if the average & median both equal 2 days, the distribution could show that 4 issues took 1 day to close, 2 issues took 2 days to close, and 4 issues took 3 days to close.

We could possibly remove the reference to median, and change the distribution items to percentiles. For example:

+ Average time of open issues 
+ Average time of first response
+ Percentile distribution of open issue time
+ Percentile distribution of issue resolution time

This seems a bit more clear to me, so what do you think?

**Comment by GeorgLink** [comment]
@BenLloydPearson 
Thanks for the immediate response.

> As I mentioned in my original post, in issue #25 we discussed moving these metrics into a different category so that these metrics can focus on the software itself, while Diversity & Participation focuses on the community side of things.

Thank you for reminding me. 
At OSSEU, we received a comment that Growth-Maturity-Decline is developer (code) centric and ignores users. Removing developers and focusing entirely on Software seems to make this distinction more pointed.
I see the value in a software centric view and a people centric view. The existing metrics are holistic and focus on questions about the community, rather then components of the community.Growth, Maturity, and Decline are not only indicated by software related metrics but also people related metrics. Diversity and Participation are not only indicated by people related metrics but also software related metrics.
Just as I like the software/people centric views, I like the question oriented and holistic metric groups. 

I don't have a solution for this challenge, nor do I know which view is "better" (more valuable/userful/intuitive). My goal here is to point out that removing the community related metrics changes the nature of the Growth, Maturity, and Decline metric groups.

Thoughts?


> Once we've finalized this PR, I was planning on expanding the Diversity section in a similar way as this.

Awesome, Thanks! Perfect time to incorporate the feedback from OSSEU.

File: 2_Growth-Maturity-Decline.md (hunk 1 above)
Comment (GeorgLink) [reply 2_Growth-Maturity-Decline.md]: Thanks for clarifying. The explanation makes sense. The proposed names are also more clear. I have a hunch that the last two could be expressed more simply - I'll sleep over it.

**Comment by BenLloydPearson** [comment]
@GeorgLink 

> I don't have a solution for this challenge, nor do I know which view is "better" (more valuable/userful/intuitive). My goal here is to point out that removing the community related metrics changes the nature of the Growth, Maturity, and Decline metric groups.

Now you're making me question other items in this PR :-)

I think the interpretation you provided makes sense, but this would mean that the Project Communication section in this PR should probably be included under Diversity and Participation rather than here. I'd have no problem with moving it over there in a separate PR after this.

File: 2_Growth-Maturity-Decline.md (hunk 6 above)
Comment (BenLloydPearson) [reply 2_Growth-Maturity-Decline.md]: Agreed, this sounds better: Average & median time for first maintainer response.

Imo, this solves two problems, 1) it focuses on responses that have meaning, and 2) it doesn't leave the door open after the final response in a thread (i.e. there won't be a way to judge if a conversation is "done" so the clock would keep ticking after the last response to a thread and would make this number meaningless).

File: 2_Growth-Maturity-Decline.md (hunk 7 above)
Comment (BenLloydPearson) [reply 2_Growth-Maturity-Decline.md]: > Average & median communication responses per thread

This metric doesn't relate to time, it counts the number of actual posts that are made in response to a thread. E.g. if two people make one post each in response to someone's email, then this number would be two.

**Comment by germonprez** [comment]
Hi, 
Late to this conversation but spent time reading through. I think I'm getting it. I do think that it makes sense to: 

- Move the communication related metrics to Participation and Inclusion (Diversity now as its own metric)

**Comment by BenLloydPearson** [comment]
I think my latest series of commits addressed all of the concerns that were brought up. Let me know if anything else needs further work.

**Comment by germonprez** [comment]
Looks great! Thanks @BenLloydPearson 

I vote for merge on this. @GeorgLink any details you see?

**Comment by jgbarah** [comment]
+1 (with respect to my comments)

**Comment by rpaik** [comment]
+1.  Looks good to me....

**Comment by bproffitt** [comment]
This looks good as a starter document.

**Comment by GeorgLink** [comment]
@BenLloydPearson  Thanks for all the work - looks great!

Everyone else, thanks for commenting and reviewing.
